### PR TITLE
Support GEOSEARCH and GEOSEARCHSTORE commands

### LINKED ISSUE #2390: GEOSEARCH/GEOSEARCHSTORE commands for bounding box spatial queries (#8094)

### PR DESCRIPTION
**Pull request by AvitalFineRedis**
Support GEOSEARCH and GEOSEARCHSTORE commands.
Also created GeoSearchParam class to support all the params for both commands.

Based on old and un-updated [PR ](https://github.com/redis/jedis/pull/2677)

Resolves #2390 
Closes #2677

### CODE REVIEW

**Review by sazzad16 — CHANGES_REQUESTED** [review]
LGTM for String variants. Please add binary (`byte[] key`) variants.

**Comment by AvitalFineRedis** [comment]
> LGTM for String variants. Please add binary (`byte[] key`) variants.

Pipeline also?

**Comment by sazzad16** [comment]
> Pipeline also?

Yes :smiley:

File: src/test/java/redis/clients/jedis/commands/jedis/GeoCommandsTest.java [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 import redis.clients.jedis.GeoCoordinate;
 import redis.clients.jedis.args.GeoUnit;
+import redis.clients.jedis.params.GeoSearchParam;
```
Comment (sazzad16) [inline src/test/java/redis/clients/jedis/commands/jedis/GeoCommandsTest.java:14]: Almost forgot. Add the new tests in `...unified...` package as well.

File: src/main/java/redis/clients/jedis/params/GeoSearchParam.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+import static redis.clients.jedis.Protocol.Keyword.BYBOX;
+import static redis.clients.jedis.Protocol.Keyword.BYRADIUS;
+import static redis.clients.jedis.Protocol.Keyword.COUNT;
+import static redis.clients.jedis.Protocol.Keyword.DESC;
+import static redis.clients.jedis.Protocol.Keyword.WITHCOORD;
+import static redis.clients.jedis.Protocol.Keyword.WITHDIST;
+import static redis.clients.jedis.Protocol.Keyword.WITHHASH;
+import static redis.clients.jedis.Protocol.Keyword.FROMMEMBER;
+import static redis.clients.jedis.Protocol.Keyword.FROMLONLAT;
+
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/params/GeoSearchParam.java:17]:
```suggestion
import static redis.clients.jedis.Protocol.Keyword.ANY;
import static redis.clients.jedis.Protocol.Keyword.ASC;
import static redis.clients.jedis.Protocol.Keyword.BYBOX;
import static redis.clients.jedis.Protocol.Keyword.BYRADIUS;
import static redis.clients.jedis.Protocol.Keyword.COUNT;
import static redis.clients.jedis.Protocol.Keyword.DESC;
import static redis.clients.jedis.Protocol.Keyword.WITHCOORD;
import static redis.clients.jedis.Protocol.Keyword.WITHDIST;
import static redis.clients.jedis.Protocol.Keyword.WITHHASH;
import static redis.clients.jedis.Protocol.Keyword.FROMMEMBER;
import static redis.clients.jedis.Protocol.Keyword.FROMLONLAT;

import redis.clients.jedis.GeoCoordinate;
import redis.clients.jedis.CommandArguments;
import redis.clients.jedis.args.GeoUnit;
```

File: src/main/java/redis/clients/jedis/params/GeoSearchParam.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            }
+        }
+
+        if (asc) {
+            args.add(ASC);
+        } else if (desc) {
+            args.add(DESC);
+        }
+    }
+}
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/params/GeoSearchParam.java:166]: Change to 2-space formatting.

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/params/GeoSearchParam.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+
+public class GeoSearchParam implements IParams {
+    private boolean frommember = false;
+    private boolean fromlonlat = false;
+    private String member;
+    private GeoCoordinate lonlat;
+
+    private boolean byradius = false;
+    private boolean bybox = false;
```
Comment (dengliming) [inline src/main/java/redis/clients/jedis/params/GeoSearchParam.java]: few suggestions : )
```
frommember => fromMember
fromlonlat  => fromLonlat  
byradius => byRadius 
bybox => byBox 
private GeoCoordinate lonlat; => private GeoCoordinate coord; 
```

File: src/main/java/redis/clients/jedis/params/GeoSearchParam.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    withDist = true;
+    return this;
+  }
+
+  public GeoSearchParam withHash() {
+    withHash = true;
+    return this;
+  }
+
+  public GeoSearchParam sortAscending() {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/params/GeoSearchParam.java:94]:
```suggestion
  public GeoSearchParam asc() {
```

File: src/main/java/redis/clients/jedis/params/GeoSearchParam.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    asc = true;
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/params/GeoSearchParam.java:95]: There are many tests with ASC. Please add more tests without any of [ASC | DESC]. Or, at least, remove ASC from some ASC tests.

File: src/main/java/redis/clients/jedis/params/GeoSearchParam.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    return this;
+  }
+
+  public GeoSearchParam sortDescending() {
+    desc = true;
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/params/GeoSearchParam.java:100]: There are no tests with DESC. Please add some. Or, at least, change few ASC tests to DESC.

File: src/main/java/redis/clients/jedis/params/GeoSearchParam.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    withHash = true;
+    return this;
+  }
+
+  public GeoSearchParam sortAscending() {
+    asc = true;
+    return this;
+  }
+
+  public GeoSearchParam sortDescending() {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/params/GeoSearchParam.java:99]:
```suggestion
  public GeoSearchParam desc() {
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/params/GeoSearchParam.java (hunk 6 above)
Comment (AvitalFineRedis) [reply src/main/java/redis/clients/jedis/params/GeoSearchParam.java:95]: There is one ASC and one DESC use :sweat_smile:

**Review by sazzad16 — APPROVED** [review]

File: src/main/java/redis/clients/jedis/params/GeoSearchParam.java [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    return this;
+  }
+
+  public GeoSearchParam asc() {
+    asc = true;
+    return this;
+  }
+
+  public GeoSearchParam desc() {
+    desc = true;
```
Comment (zeekling) [inline src/main/java/redis/clients/jedis/params/GeoSearchParam.java:100]: can merge desc and asc into one param?

File: src/test/java/redis/clients/jedis/commands/jedis/GeoCommandsTest.java [hunk 10]
```diff
@@ -470,6 +471,139 @@ public void georadiusByMemberReadonlyBinary() {
...
+    members = jedis.geosearch("barcelona", new GeoSearchParam().byBox(1,1, GeoUnit.KM)
+            .fromLonLat(2.191, 41.433).withDist().withCoord());
+    assertEquals(1, members.size());
+    assertEquals("place1", members.get(0).getMemberByString());
+    assertEquals(0.0881, members.get(0).getDistance(), 10);
+    assertEquals(new GeoCoordinate(2.19093829393386841, 41.43379028184083523), members.get(0).getCoordinate());
+  }
+
+  @Test
+  public void geosearch_negative() {
```
Comment (zeekling) [inline src/test/java/redis/clients/jedis/commands/jedis/GeoCommandsTest.java:531]:
```suggestion
  public void geosearchNegative() {
```

**Review by sazzad16 — APPROVED** [review]

**Review by sazzad16 — CHANGES_REQUESTED** [review]
@AvitalFineRedis Undo any change related to GeoRadiusParam.

File: src/main/java/redis/clients/jedis/params/GeoSearchParam.java [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+    if (count != null) {
+      args.add(COUNT).add(count);
+      if (any) {
+        args.add(ANY);
+      }
+    }
+
+    if (asc) {
+      args.add(ASC);
+    } else {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/params/GeoSearchParam.java:161]: else if (desc)

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/params/GeoSearchParam.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+    if (count != null) {
+      args.add(COUNT).add(count);
+      if (any) {
+        args.add(ANY);
+      }
+    }
+
+    if (asc) {
+      args.add(ASC);
+    } else if (desc){
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/params/GeoSearchParam.java:162]:
```suggestion
    } else if (desc) {
```

File: src/main/java/redis/clients/jedis/params/GeoSearchParam.java [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    return this;
+  }
+
+  public GeoSearchParam desc() {
+    desc = true;
+    return this;
+  }
+
+  public GeoSearchParam count(int count) {
+    if (count > 0) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/params/GeoSearchParam.java:105]: return count(count, false)

**Review by sazzad16 — CHANGES_REQUESTED** [review]

**Review by sazzad16 — APPROVED** [review]

**Review by dengliming — APPROVED** [review]

**Pull request merged**